### ISSUE TITLE
Allow multiple application contexts per app

### ISSUE THREAD

**Issue by benfrancis** [w3c/manifest#294]
Currently it appears that an application is limited to a single application context per app. The application context is created when the app is launched, calls to window.open and hyperlinks with target=_blank can only open a browsing context, navigations to URLs within the scope of an app inside a browsing context will never result in an application context being created.

As an app developer I would like to be able to create multiple application contexts for a web app in the same way that I can create multiple browsing contexts for a web site. When I call window.open() or use a hyperlink with target=_blank to a URL within the scope of the app, I would like the newly created browsing context to be an application context.

In other words, I would like multiple windows per app.

**Comment by benfrancis** [comment]
Chrome OS supports multiple windows per app https://developer.chrome.com/apps/app_lifecycle#create_windows so the Chrome Web Store might be a good place to look for existing use cases. We're unlikely to find use cases in native mobile apps because it simply isn't possible.

Given that a browsing context can create other browsing contexts and an application context is a type of browsing context, I would re-phrase the question as "why should application contexts be artificially limited to one per app?". I would argue that just because that's a limitation of existing mobile platforms doesn't mean that limitation should be carried over to the web.

**Comment by marcoscaceres** [comment]
> why should application contexts be artificially limited to one per app?

It's not artificial until we find valid use cases (and use cases that are not covered by window.open). That's not to say there are no good use cases - but we should find those first and see who actually needs this (particularly people who are limited by window.open) ... and then, and only then, think about adding something like this.

**Comment by marcoscaceres** [comment]
@benfrancis wrote: 
> the Chrome Web Store might be a good place to look for existing use cases.

So, [looking at why Chrome apps seem to need this](https://github.com/search?utf8=%E2%9C%93&q=%22chrome.app.window.create%22&type=Code&ref=searchresults
)... it appears that they are used by background and event pages (`background.js`, etc.):

Because js-based background and event pages have no DOM or representation, it makes sense to sometimes pop-up a window to allow the user to do something (using HTML). 

If we were to make some kind of analogy, I guess we could view a background/event page as a service worker. If that is the case, then the service worker may be able to create new browsing contexts of some type and in some way (independently of whatever the manifest spec says) - like, an auxiliary browsing context created with `window.open()`. As chrome's background.js things don't have a DOM, they don't have `window.open` and hence can't rely on it. It will probably be the same case with Service Workers - but that's best taken up with the service worker folks. 

Further, [Chrome's documentation](https://developer.chrome.com/apps/app_lifecycle#create_windows) states that: "By default, these windows are created with a script connection to the event page and are directly scriptable by the event page." Which implies some kind of parent-child relationship (not a new independent application context). 

I propose we close this until we see what all the limitations are with just using `window.open`. Then we can reopen once we have some concrete use cases and demand from developers (that can't be addressed in a sane way with `window.open`).

**Comment by marcoscaceres** [comment]
Closing for now. Will return to this once we see what people start making.

**Comment by benfrancis** [comment]
> It's not artificial until we find valid use cases (and use cases that are not covered by window.open). That's not to say there are no good use cases - but we should find those first and see who actually needs this (particularly people who are limited by window.open) ... and then, and only then, think about adding something like this.

Fair enough. So to confirm, as the spec is currently written calling window.open() from an application context will always create a browsing context without the manifest applied? The only way to create an application context with the manifest applied is by launching an app?

Just for refrerence, here is the algorithm from Jonas' original application scopes proposal, which I agree with:

>When opening a new context:
>1. If the new context is a "dialog" context, create a new dialog
>window and open the URL there.
>2. If the "target" of the navigation is "_blank" and the URL is not
>part of the current context's Apps' "scope", go to step 4.
>3. If the URL is part of the current context's App's "scope" or
>"open-in-app", open the new page within the current app.
>4. If the URL is part of multiple other App's "scope", show an
>application picker to the user and allow the user to choose which of
>the two applications to use.
>5. If the URL is part of another App's "scope", open the new page
>within that App.
>6. Open the URL in the system browser.
https://groups.google.com/forum/#!topic/mozilla.dev.b2g/Ufrz9UW-Z28

The reason I think this is important is to allow multiple pages of an app to be open at the same time, with the manifest applied. This becomes particularly important if we add capturing scope to the spec such that an app can capture navigations from a browsing context outside the app. Otherwise any navigation to a URL inside the scope of an app would always navigate the single application context for that app, potentially navigating the user away from a page they are still using.

**Comment by marcoscaceres** [comment]
>  Fair enough. So to confirm, as the spec is currently written calling window.open() from an application context will always create a browsing context without the manifest applied? 

correct. 

> The only way to create an application context with the manifest applied is by launching an app?

Correct. This could also help to mitigate a kind of masquerading attack: as opening an app with window.open can create a clear overlay with browser chrome.  

> Just for reference, here is the algorithm from Jonas' original application scopes proposal, which I agree with:

I think parts of Jonas' proposal is very OS specific - it makes UI assumptions (e.g., showing a chooser). For instance, it wouldn't make sense in iOS - where these kinds of navigation are handled automagically (they get routed through Safari, which is pretty funny to watch). However, it would be fine to use that algorithm in FxOS. 

The way that scope is currently specified allows implementation to freely adapt the concept to their own needs and underlying OS conventions. I think we should let implementers play around with the concept of scope and see what emerges. If we see that there is something to standardize, then we can get agreement and do that.

**Comment by marcoscaceres** [comment]
> The reason I think this is important is to allow multiple pages of an app to be open at the same time, with the manifest applied. This becomes particularly important if we add capturing scope to the spec such that an app can capture navigations from a browsing context outside the app. Otherwise any navigation to a URL inside the scope of an app would always navigate the single application context for that app, potentially navigating the user away from a page they are still using. 

Can you walk through the above sequentially from the following scenarios below? I want to make sure we are on the same page (hehe, no pun intended). 

We assume the following two apps are installed:

### App 1 - Foo!
  * short name: "Foo!" 
  * scope: is foo.com/app.
  * display mode: standalone.
  * orientation: portrait. 
  * start url: foo.com/app/start.html

### App 2 - Bar! 
  * short name: "Bar!" 
  * scope: is bar.com/.
  * display mode: fullscreen.
  * orientation: landscape. 
  * start url: bar.com/start.html

### Installed app (in app navigation within scope)
1. User launches "Foo!" by clicking its icon. 
1. App context automatically navigates to `foo.com/app/start.html`.
1. User clicks link to `foo.com/app/whatever.html`. Link is in scope. 
1. Navigation to `whatever.html` happens and application context remains fullscreen.
 
### Browser navigation
1. User has opened Foo!. 
1. User navigates Foo! to "foo.com/app/form.html" - inputs some details. 
1. User jumps to web browser and navigates to facebook.com. This puts Foo! in the background. 
1. In browser, user clicks on link to `foo.com/app`. 
1. Browser navigates to `foo.com/app`. 
1. As foo.com/app is in scope of an installed web app (Foo!), browser MAY indicate that Foo! is already installed and the user can switch to Foo app: 
![screenshot 2015-01-12 11 02 04](https://cloud.githubusercontent.com/assets/870154/5698471/3447e1c2-9a5a-11e4-97d4-dae6e1caf7ec.png)

1. In browser, user navigates to foo.com/app/whatever.html (also in scope).
1. User clicks on "open" button, Foo app opens `foo.com/app/whatever.html`. 
1. foo.com/form.html detects that it's about to `unload`. Alerts users that they could lose work.  

### Installed app (following links outside scope)
1. user opens Foo! 
1. User clicks on `bbc.co.uk/news`. Link is out of scope. 
1. Link opens in default web browser. 

### Installed app (following links)
1. user opens Foo! 
1. User clicks on `bbc.co.uk/news` with `_blank` (or `_whatever`). 
1. Link opens in browser. 

### Installed app (opening windows)
1. user opens Foo! 
1. User clicks on `bbc.co.uk/news/some_story`. 
1. Click is intercepted, and `window.open` is called.   
1. Overlay browser is shown. `bbc.co.uk/news/some_story` is shown. 
1. User closes overlay browser, returns to app (currently `at foo.com/app/start.html`).  

### Jumping between Bar and Foo
1. user opens Foo! 
1. User clicks on `bar.com/whatever`.
1. Open in Bar! Orientation and display mode change. 
1. User clicks "Return to Foo!" (foo.com/app/whatever).
1. Open in Foo! Orientation and display mode change.

### Notes
Identity is still problematic. There can be a third app that can subsume Foo!'s scope and start URL. 
Need to deal with that in a separate bug.

**Comment by benfrancis** [comment]
> This could also help to mitigate a kind of masquerading attack: as opening an app with window.open can create a clear overlay with browser chrome. 

Can you explain this further? I'm not sure I understand what you mean.

> I think parts of Jonas' proposal is very OS specific

Sorry, perhaps I should have explicitly pointed out that the quote was taken from a proposal specific to Firefox OS, not for this W3C specification, it was intentionally OS-specific. It was the general algorithm I was citing, not the specific UI details.

> The way that scope is currently specified allows implementation to freely adapt the concept to their own needs and underlying OS conventions. I think we should let implementers play around with the concept of scope and see what emerges. If we see that there is something to standardize, then we can get agreement and do that.

That's fair enough. An Android/iOS/Windows/OS X etc. implementation may not want to implement the capturing scope part. Waiting to see what implementers do sounds reasonable as deep linking into apps is a new area. There are certainly lots of proprietary non-web approaches to deep linking being explored at the moment, maybe one of those will supersede the web instead (I hope not).

**Comment by benfrancis** [comment]
> Installed app (in app navigation within scope)

Looks fine. (Terminology nit: Remains "standalone" without browser chrome, not "fullscreen").

> Browser navigation
...
> 6. As foo.com/app is in scope of an installed web app (Foo!), browser MAY indicate that Foo! is already installed and the user can switch to Foo app:

I would actually go further than this and say that the user agent MAY just automatically switch to the app and carry out the navigation there, rather than prompting the user. Once an app is installed, it is registered as managing that part of the web, and should always do so. But that should probably be left down to the implementation.

> 9. foo.com/form.html detects that it's about to unload. Alerts users that they could lose work.

If an app can only have one application context, this would be the only option. If an app could have multiple application contexts, the user agent could instead create an additional application context for the app to navigate to foo.com/app/whatever.html without having to unload foo.com/form.html.

> Installed app (following links outside scope)

Looks fine.

> Installed app (opening windows)
...
> 4. Overlay browser is shown. bbc.co.uk/news/some_story is shown. 

I think that by default this should open in a new browsing context in the browser, not an overlay window on top of the app, because it is out of scope of the app. There may be cases where the app author wants to opt-in to staying inside the app, for that I suggest something like using a "dialog" feature of window.open().

> Jumping between Bar and Foo

Looks fine.

> Identity is still problematic. There can be a third app that can subsume Foo!'s scope and start URL.
Need to deal with that in a separate bug.

I think two apps shouldn't be allowed to have overlapping scope. This could be done by either by the user agent refusing to install an app with a scope which overlaps that of another installed app, or alternatively by using the "last registration wins" approach taken by Service Worker scope.

**Comment by marcoscaceres** [comment]
On January 12, 2015 at 11:05:44 PM, Ben Francis (notifications@github.com) wrote:
> > 6. As foo.com/app is in scope of an installed web app (Foo!), browser MAY indicate that  
> Foo! is already installed and the user can switch to Foo app:
>  
> I would actually go further than this and say that the user agent MAY just automatically  
> switch to the app and carry out the navigation there, rather than prompting the user.  
> Once an app is installed, it is registered as managing that part of the web, and should  
> always do so. But that should probably be left down to the implementation.

Agree. That's a UX decision. I personally wouldn't like this, as when I'm "surfing the web" I generally don't want to be taken out of that context. 

> > 9. foo.com/form.html detects that it's about to unload. Alerts users that they could  
> lose work.
>  
> If an app can only have one application context, this would be the only option. If an app  
> could have multiple application contexts, the user agent could instead create an additional  
> application context for the app to navigate to foo.com/app/whatever.html without  
> having to unload foo.com/form.html.

Again, this is something for UX people to play around with. IMO, it could become really confusing if multiple versions of an app were running at the same time: thinking of how confusing it is having multiple application windows in traditional desktop software and how easy it is to lose a window. 

> > Installed app (opening windows)
> ...
> > 4. Overlay browser is shown. bbc.co.uk/news/some_story is shown.
>  
> I think that by default this should open in a new browsing context in the browser, not an  
> overlay window on top of the app, because it is out of scope of the app.

Again, it's a UX thing to explore. For me, jumping to another app breaks user expectation - in that it yanks the user from one context into another context (thinking of how confusing Android is in this respect). The overlay model is, IMO, better because it doesn't take the user from one application context to another: through the appropriate visual queues, it's always clear which application you are in and what you are doing: user story - "I'm in twitter, I followed a link to a news story. I'm reading the story, following links within that story... and if I press here (X) I close this overlay window and return to twitter". 

> There may be cases  
> where the app author wants to opt-in to staying inside the app, for that I suggest something  
> like using a "dialog" feature of window.open().

We would need to look at individual cases. The only use case I can think of for leaving an app is to have another application do something as part of an action (or because the overlay is not able to render the page properly for whatever reason - and I need to open the page in a different browser). 

For general browsing, can you think of any reason why one would want to open a browser from within an application without using an overlay browsing context? 

> > Identity is still problematic. There can be a third app that can subsume Foo!'s scope  
> and start URL.
> Need to deal with that in a separate bug.
>  
> I think two apps shouldn't be allowed to have overlapping scope. This could be done by  
> either by the user agent refusing to install an app with a scope which overlaps that of  
> another installed app, or alternatively by using the "last registration wins" approach  
> taken by Service Worker scope.

I was thinking something similar - however, both approaches have really bad consequences, which directly affect users, rather than applications (which is why I think it's ok for service workers to have the "last reg. wins" model, but not ok for manifest). 

Anyway, let's handle that elsewhere.

**Comment by benfrancis** [comment]
> For general browsing, can you think of any reason why one would want to open a browser from within an application without using an overlay browsing context?

Try using the Android Facebook app. By default it will open external links in a very basic overlay on top of the app rather than respect my choice and use my fully featured default browser. In my opinion, if I click on a hyperlink which doesn't fall within the scope of an installed app, it should open as a new tab in my browser, not stay inside the context of the app with some kind of stripped down overlay. Once I've left the scope of the app I'm back out on the open web and free to follow further hyperlinks, it doesn't make sense to keep me in the app.

These are largely UX decisions specific to a particular user agent so as long as the spec doesn't prevent an implementation from making these kinds of choices, I think that's fine. I think defining an application context as a browsing context to which a manifest has been applied is fine. I think specifying that an application context can only be created by launching an app would be overly restrictive (I don't think the spec currently says this). I think saying that an application context MUST always be navigated to the start_url when it is created may be too strong, the user agent might want to deep link into an app and therefore navigate a new application context to a URL other than the default start URL used when launching the app.

So perhaps the sentence "When an application context is created, the user agent MUST immediately navigate to the start URL with replacement enabled" is a little strong?

**Comment by marcoscaceres** [comment]
> Try using the Android Facebook app. By default it will open external links in a very basic 
> overlay on top of the app rather than respect my choice and use my fully featured default 
> browser. 

This is generally a business decision. You don't want the user to leave your app - but you want them to complete the task they need to complete (e.g., reading a news article, viewing a funny picture, etc.). Thus this need a good balance between what developers want and what users wants.  

> In my opinion, if I click on a hyperlink which doesn't fall within the scope of 
> an installed app, it should open as a new tab in my browser, not stay inside the context 
> of the app with some kind of stripped down overlay.

The overlay might provide that choice to the user, as Twitter on iOS does: 
![screenshot_2015-01-14_10_14_15](https://cloud.githubusercontent.com/assets/870154/5731979/5adacab6-9bd6-11e4-9a66-1e583d56fd88.png)
 
The thing is to provide enough choice to satisfy everyone.  

> Once I've left the scope of the app I'm 
> back out on the open web and free to follow further hyperlinks, it doesn't make sense to 
> keep me in the app.

I strongly disagree. Jumping from one app to another to view a web page would be tremendously annoying to most users. It's why this (stripped down browsers) is such a prevalent pattern in so many apps. They allow users to do whatever they need to do (view an article, watch a quick video, etc.) without actually having to go to the big heavy browser. It avoids a big context switch, it's fast, etc. and keeps users focused.   

> These are largely UX decisions specific to a particular user agent so as long as the spec 
> doesn't prevent an implementation from making these kinds of choices, I think that's 
> fine.

Agree :+1: 

>  I think defining an application context as a browsing context to which a manifest 
> has been applied is fine. I think specifying that an application context can only be created 
> by launching an app would be overly restrictive (I don't think the spec currently says 
> this). 

Yeah, the spec doesn't prevent this - and I think it would be awesome for implementers to experiment with this. 

> I think saying that an application context MUST always be navigated to the start_url 
> when it is created may be too strong,

Ok, yes - I think I see what you mean. The text was written from the perspective of launching the application from the homescreen - not from a "deep link". It was certainly not the intention to prevent "deep linking". I'll clarify. 

>  the user agent might want to deep link into an app and 
> therefore navigate a new application context to a URL other than the default start URL 
> used when launching the app.

Agree. 

> So perhaps the sentence "When an application context is created, the user agent MUST 
> immediately navigate to the start URL with replacement enabled" is a little strong? 

I'll fix that up.
